### ISSUE TITLE
Package request: Advanced Copy (advcpmv)

### ISSUE THREAD

**Issue by amalgame21** [NixOS/nixpkgs#271508]
**Project description**

'cp' and 'mv' utilities with progress bar patches
The [nnn terminal file manager](https://github.com/jarun/nnn) utilize it (flag -r) to show cp, mv progress.

Advanced Copy is a mod for the GNU cp and GNU mv tools which adds a progress bar and provides some info on what's going on. It was written by Florian Zwicke and released under the GPL.

This repository exists because the [Advanced Copy](http://beatex.org/web/advancedcopy.html) website appears to be dead. You can still find it via the [Internet Archive](https://web.archive.org/web/20131115171331/http://beatex.org/web/advancedcopy.html).

More info: https://aur.archlinux.org/packages/advcpmv

**Metadata**

* homepage URL: https://github.com/jarun/advcpmv
* source URL: https://github.com/jarun/advcpmv
* license: GPL
* platforms: unix, linux

**Priorities**

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by amalgame21** [comment]
Yes

**Comment by AndersonTorres** [comment]
On a hand, it can simply be treated as a `overrideAttrs`, or even better, a NUR-like overlay.

On the other hand, coreutils is a core (pun not intended) package for Nixpkgs, a dependency for things from fetchers to stdenvs.
Modifying it has the potential to be a mass-rebuiding mess.
Indeed, patches for coreutils should be "vendored".

Let's try.

**Comment by amalgame21** [comment]
Sorry I am so new to Nix and NixOS (2 weeks usage), I don't even know it would mess things up.
I don't know much about `overrideAttrs`, isn't it just permanently modify `cp` and `mv` in place?
I have learned something about writing a derivation this week, and here is the .nix file I am using and work just fine for me:
```nix
with import <nixpkgs> {}; # bring all of Nixpkgs into scope

stdenv.mkDerivation rec {
	name = "advcpmv-${version}";
	version = "0.9";
	ADVCPMV_VERSION = "${version}";
	CORE_UTILS_VERSION = "9.4" ;
	src = builtins.fetchTarball {
		url = "http://ftp.gnu.org/gnu/coreutils/coreutils-${CORE_UTILS_VERSION}.tar.xz";
		sha256 = sha256:013xx0qvc20d286gfyiibqgjdlpsidyxvaq4dd4lmzvihmwd15fb;
	};
	patches = [
		(fetchpatch {
		 url = "https://raw.githubusercontent.com/jarun/advcpmv/master/advcpmv-${ADVCPMV_VERSION}-${CORE_UTILS_VERSION}.patch";
		 hash = "sha256-4fdqpkENPfra4nFQU4+xNrlfq6Dw/2JIZXUOMmdMtcM=";
		 })
	];
	buildPhase = ''
		./configure
		make
		'';
	installPhase = ''
		mkdir -p $out/bin
		cp ./src/cp $out/bin/advcp
		cp ./src/mv $out/bin/advmv
		ln -s $out/bin/advcp $out/bin/acp
		ln -s $out/bin/advcp $out/bin/cpg
		ln -s $out/bin/advmv $out/bin/amv
		ln -s $out/bin/advmv $out/bin/mvg
		'';
}
```

**Comment by AndersonTorres** [comment]
Nix is source-based - it builds the binaries (and artifacts in general) from the sources.
Think like picking Archlinux but from the PKGBUILDS.

Despite being source-based, Nix can rely on local and remote binary caches to speedup the building.

Certainly, when a package source is updated, its old entry in the cache is invalidated and the new version is put in its place.

For an application like, say, Zsnes emulator, it affects only Zsnes.

On the other, for a library like, say, SDL, its update invalidates SDL and everything that uses SDL - zsnes, bochs, frets-on-fire, dosbox, mednafen... the list is huge.

And it applies to everything, from the core utilities like coreutils to desktops like XFCE.

[XKCD 2347](https://xkcd.com/2347/) provides a better visual analogy.

**Comment by AndersonTorres** [comment]
I have tried this and it worked:

```
advcpmv = (coreutils.override { singleBinary = false; }).overrideAttrs (old:
  let
    advcpmv-data = {
      pname = "advcpmv";
      patch-version = "0.9";
      coreutils-version = "9.3";
      version = "${advcpmv-data.patch-version}-${advcpmv-data.coreutils-version}";
      src = pkgs.fetchFromGitHub {
        owner = "jarun";
        repo = "advcpmv";
        rev = "a1f8b505e691737db2f7f2b96275802c45f65c59";
        hash = "sha256-IHfMu6PyGRPc87J/hbxMUdosmLq13K0oWa5fPLWKOvo=";
      };
      patch-file = advcpmv-data.src + "/advcpmv-${advcpmv-data.version}.patch";
    };
  in assert (advcpmv-data.coreutils-version == old.version); {
    inherit (advcpmv-data) pname version;

    patches = (old.patches or [ ]) ++ [
      advcpmv-data.patch-file
    ];

    configureFlags = (old.configureFlags or [ ]) ++  [
      "--program-prefix=adv"
    ];

    postInstall = (old.postInstall or [ ]) + ''
      pushd $out/bin
      ln -s advcp cpg
      ln -s advmv mvg
      popd
    '';
    meta = old.meta // {
      description = "Coreutils patched to add progress bars";
    };
  }
);
```

**Comment by Tungsten842** [comment]
I think that this issue should stay open because it is not necessary to edit coreutils to package this. One can just make another package that only includes advcpmv commands like arch does: https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=advcpmv

**Comment by geekodour** [comment]
I just tried to rebuild my config, got:

```
         at /nix/store/b58jld7i8gjic1nrg1p01zj0k3n18139-nixos-23.11/nixos/lib/modules.nix:807:9:

          806|     in warnDeprecation opt //
          807|       { value = builtins.addErrorContext "while evaluating the option `${showOption loc}':" value;
             |         ^
          808|         inherit (res.defsFinal') highestPrio;

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: 'advcpmv' has been removed, as it is not being actively maintained and break recent coreutils.

 ```
 
 What's the suggested resolution here?

**Comment by geekodour** [comment]
https://github.com/NixOS/nixpkgs/pull/300043

From this it seems like, the only way out is to not use this package anymore? It's an dependency to some nnn file browser feature. I'll look for alternatives for it.

**Comment by AndersonTorres** [comment]
Two weeks ago [a new patch was included, targeting Coreutils 9.5](https://github.com/jarun/advcpmv/commit/1e2b1c6b74fa0974896bf94604279a3f74b37a63).

If you are interested in keeping advcpmv - and receiving some random scolding because it will break the world in a future update of Coreutils -, feel free to pick it.
